File: languages/php/Bloock/Entity/IdentityV2/SchemaBuilder.php
<?php

namespace Bloock\Entity\IdentityV2;

use Bloock\Bridge\Bridge;
use Bloock\BuildSchemaRequestV2;
use Bloock\ConfigData;
use Exception;
use Google\Protobuf\Internal\RepeatedField;

class SchemaBuilder
{
    private string $displayName;
    private string $schemaType;
    private string $version;
    private string $description;
    private string $issuerDid;
    private ConfigData $configData;

    private array $stringAttributes;
    private array $integerAttributes;
    private array $decimalAttributes;
    private array $booleanAttributes;
    private array $dateAttributes;
    private array $datetimeAttributes;
    private array $stringEnumAttributes;
    private array $integerEnumAttributes;
    private array $decimalEnumAttributes;

    public function __construct(string $displayName, string $schemaType, string $version, string $description, string $issuerDid, ConfigData $configData)
    {
        $this->displayName = $displayName;
        $this->schemaType = $schemaType;
        $this->version = $version;
        $this->description = $description;
        $this->issuerDid = $issuerDid;
        $this->configData = $configData;

        $this->stringAttributes = [];
        $this->integerAttributes = [];
        $this->decimalAttributes = [];
        $this->booleanAttributes = [];
        $this->dateAttributes = [];
        $this->datetimeAttributes = [];
        $this->stringEnumAttributes = [];
        $this->integerEnumAttributes = [];
        $this->decimalEnumAttributes = [];
    }

    public function addStringAttribute(string $name, string $technicalName, ?string $description, bool $required): SchemaBuilder
    {
        $attribute = new StringAttributeDescriptor($name, $technicalName, $description, $required);
        $this->stringAttributes[] = $attribute->toProto();
        return $this;
    }

    public function addIntegerAttribute(string $name, string $technicalName, ?string $description, bool $required): SchemaBuilder
    {
        $attribute = new IntegerAttributeDescriptor($name, $technicalName, $description, $required);
        $this->integerAttributes[] = $attribute->toProto();
        return $this;
    }

    public function addDecimalAttribute(string $name, string $technicalName, ?string $description, bool $required): SchemaBuilder
    {
        $attribute = new DecimalAttributeDescriptor($name, $technicalName, $description, $required);
        $this->decimalAttributes[] = $attribute->toProto();
        return $this;
    }

    public function addBooleanAttribute(string $name, string $technicalName, ?string $description, bool $required): SchemaBuilder
    {
        $attribute = new BooleanAttributeDescriptor($name, $technicalName, $description, $required);
        $this->booleanAttributes[] = $attribute->toProto();
        return $this;
    }

    public function addDateAttribute(string $name, string $technicalName, ?string $description, bool $required): SchemaBuilder
    {
        $attribute = new DateAttributeDescriptor($name, $technicalName, $description, $required);
        $this->dateAttributes[] = $attribute->toProto();
        return $this;
    }

    public function addDateTimeAttribute(string $name, string $technicalName, ?string $description, bool $required): SchemaBuilder
    {
        $attribute = new DatetimeAttributeDescriptor($name, $technicalName, $description, $required);
        $this->datetimeAttributes[] = $attribute->toProto();
        return $this;
    }

    public function addStringEnumAttribute(string $name, string $technicalName, ?string $description, bool $required, array $enumeration): SchemaBuilder
    {
        $repeatedField = new RepeatedField(\Google\Protobuf\Internal\GPBType::STRING);
        foreach ($enumeration as $value) {
            $repeatedField[] = $value;
        }
        $attribute = new StringEnumAttributeDescriptor($name, $technicalName, $description, $required, $repeatedField);
        $this->stringAttributes[] = $attribute->toProto();
        return $this;
    }

    public function addIntegerEnumAttribute(string $name, string $technicalName, ?string $description, bool $required, array $enumeration): SchemaBuilder
    {
        $repeatedField = new RepeatedField(\Google\Protobuf\Internal\GPBType::INT64);
        foreach ($enumeration as $value) {
            $repeatedField[] = $value;
        }
        $attribute = new IntegerEnumAttributeDescriptor($name, $technicalName, $description, $required, $repeatedField);
        $this->stringAttributes[] = $attribute->toProto();
        return $this;
    }

    public function addDecimalEnumAttribute(string $name, string $technicalName, ?string $description, bool $required, array $enumeration): SchemaBuilder
    {
        $repeatedField = new RepeatedField(\Google\Protobuf\Internal\GPBType::DOUBLE);
        foreach ($enumeration as $value) {
            $repeatedField[] = $value;
        }
        $attribute = new DecimalEnumAttributeDescriptor($name, $technicalName, $description, $required, $repeatedField);
        $this->stringAttributes[] = $attribute->toProto();
        return $this;
    }

    public function build(): Schema
    {
        $bridge = new Bridge();

        $req = new BuildSchemaRequestV2();
        $req->setConfigData($this->configData);
        $req->setDisplayName($this->displayName);
        $req->setSchemaType($this->schemaType);
        $req->setVersion($this->version);
        $req->setDescription($this->description);
        $req->setIssuerDid($this->issuerDid);
        $req->setStringAttributes($this->stringAttributes);
        $req->setIntegerAttributes($this->integerAttributes);
        $req->setDecimalAttributes($this->decimalAttributes);
        $req->setBooleanAttributes($this->booleanAttributes);
        $req->setDateAttributes($this->dateAttributes);
        $req->setDatetimeAttributes($this->datetimeAttributes);
        $req->setStringEnumAttributes($this->stringEnumAttributes);
        $req->setIntegerEnumAttributes($this->integerEnumAttributes);
        $req->setDecimalEnumAttributes($this->decimalEnumAttributes);

        $res = $bridge->identityV2->BuildSchema($req);

        if ($res->getError() != null) {
            throw new Exception($res->getError()->getMessage());
        }

        return Schema::fromProto($res->getSchema());
    }
}

File: languages/java/src/main/java/com/bloock/sdk/entity/identity_v2/SchemaBuilder.java
package com.bloock.sdk.entity.identity_v2;

import com.bloock.sdk.bridge.Bridge;
import com.bloock.sdk.bridge.proto.Config;
import com.bloock.sdk.bridge.proto.IdentityEntitiesV2;
import com.bloock.sdk.bridge.proto.IdentityV2;
import com.bloock.sdk.bridge.proto.Shared;
import java.util.ArrayList;
import java.util.List;

/**
 * Is a builder pattern for constructing schema instances.
 */
public class SchemaBuilder {
  private final String displayName;
  private final String schemaType;
  private final String version;
  private final String description;
  private final Config.ConfigData configData;

  private final List<IdentityEntitiesV2.StringAttributeDefinitionV2> stringAttributes;
  private final List<IdentityEntitiesV2.IntegerAttributeDefinitionV2> integerAttributes;
  private final List<IdentityEntitiesV2.DecimalAttributeDefinitionV2> decimalAttributes;
  private final List<IdentityEntitiesV2.BooleanAttributeDefinitionV2> booleanAttributes;
  private final List<IdentityEntitiesV2.DateAttributeDefinitionV2> dateAttributes;
  private final List<IdentityEntitiesV2.DateTimeAttributeDefinitionV2> datetimeAttributes;
  private final List<IdentityEntitiesV2.StringEnumAttributeDefinitionV2> stringEnumAttributes;
  private final List<IdentityEntitiesV2.IntegerEnumAttributeDefinitionV2> integerEnumAttributes;
  private final List<IdentityEntitiesV2.DecimalEnumAttributeDefinitionV2> decimalEnumAttributes;

  /**
   * Creates a new instance of SchemaBuilder with initial values.
   * @param displayName
   * @param schemaType
   * @param version
   * @param description
   * @param configData
   */
  public SchemaBuilder(
      String displayName,
      String schemaType,
      String version,
      String description,
      Config.ConfigData configData) {
    this.displayName = displayName;
    this.schemaType = schemaType;
    this.version = version;
    this.description = description;
    this.configData = configData;

    this.stringAttributes = new ArrayList<>();
    this.integerAttributes = new ArrayList<>();
    this.decimalAttributes = new ArrayList<>();
    this.booleanAttributes = new ArrayList<>();
    this.dateAttributes = new ArrayList<>();
    this.datetimeAttributes = new ArrayList<>();
    this.stringEnumAttributes = new ArrayList<>();
    this.integerEnumAttributes = new ArrayList<>();
    this.decimalEnumAttributes = new ArrayList<>();
  }

  /**
   * Adds a string attribute descriptor to the schema builder.
   * @param name
   * @param technicalName
   * @param description
   * @param required
   * @return
   */
  public SchemaBuilder addStringAttribute(
      String name, String technicalName, String description, Boolean required) {
    this.stringAttributes.add(
        new StringAttributeDescriptor(name, technicalName, description, required).toProto());
    return this;
  }

  /**
   * Adds an integer attribute descriptor to the schema builder.
   * @param name
   * @param technicalName
   * @param description
   * @param required
   * @return
   */
  public SchemaBuilder addIntegerAttribute(
      String name, String technicalName, String description, Boolean required) {
    this.integerAttributes.add(
        new IntegerAttributeDescriptor(name, technicalName, description, required).toProto());
    return this;
  }

  /**
   * Adds a decimal attribute descriptor to the schema builder.
   * @param name
   * @param technicalName
   * @param description
   * @param required
   * @return
   */
  public SchemaBuilder addDecimalAttribute(
      String name, String technicalName, String description, Boolean required) {
    this.decimalAttributes.add(
        new DecimalAttributeDescriptor(name, technicalName, description, required).toProto());
    return this;
  }

  /**
   * Adds a boolean attribute descriptor to the schema builder.
   * @param name
   * @param technicalName
   * @param description
   * @param required
   * @return
   */
  public SchemaBuilder addBooleanAttribute(
      String name, String technicalName, String description, Boolean required) {
    this.booleanAttributes.add(
        new BooleanAttributeDescriptor(name, technicalName, description, required).toProto());
    return this;
  }

  /**
   * Adds a date attribute descriptor to the schema builder.
   * @param name
   * @param technicalName
   * @param description
   * @param required
   * @return
   */
  public SchemaBuilder addDateAttribute(
      String name, String technicalName, String description, Boolean required) {
    this.dateAttributes.add(
        new DateAttributeDescriptor(name, technicalName, description, required).toProto());
    return this;
  }

  /**
   * Adds a datetime attribute descriptor to the schema builder.
   * @param name
   * @param technicalName
   * @param description
   * @param required
   * @return
   */
  public SchemaBuilder addDatetimeAttribute(
      String name, String technicalName, String description, Boolean required) {
    this.datetimeAttributes.add(
        new DatetimeAttributeDescriptor(name, technicalName, description, required).toProto());
    return this;
  }

  /**
   * Adds a string enum attribute descriptor to the schema builder.
   * @param name
   * @param technicalName
   * @param description
   * @param required
   * @param enumeration
   * @return
   */
  public SchemaBuilder addStringEnumAttribute(
      String name,
      String technicalName,
      String description,
      Boolean required,
      List<String> enumeration) {
    this.stringEnumAttributes.add(
        new StringEnumAttributeDescriptor(name, technicalName, description, required, enumeration)
            .toProto());
    return this;
  }

  /**
   * Adds an integer enum attribute descriptor to the schema builder.
   * @param name
   * @param technicalName
   * @param description
   * @param required
   * @param enumeration
   * @return
   */
  public SchemaBuilder addIntegerEnumAttribute(
      String name,
      String technicalName,
      String description,
      Boolean required,
      List<Long> enumeration) {
    this.integerEnumAttributes.add(
        new IntegerEnumAttributeDescriptor(name, technicalName, description, required, enumeration)
            .toProto());
    return this;
  }

  /**
   * Adds a decimal enum attribute descriptor to the schema builder.
   * @param name
   * @param technicalName
   * @param description
   * @param required
   * @param enumeration
   * @return
   */
  public SchemaBuilder addDecimalEnumAttribute(
      String name,
      String technicalName,
      String description,
      Boolean required,
      List<Double> enumeration) {
    this.decimalEnumAttributes.add(
        new DecimalEnumAttributeDescriptor(name, technicalName, description, required, enumeration)
            .toProto());
    return this;
  }

  /**
   * Creates a schema using the configured attributes.
   * @return
   * @throws Exception
   */
  public Schema build() throws Exception {
    Bridge bridge = new Bridge();

    IdentityV2.BuildSchemaRequestV2 req =
        IdentityV2.BuildSchemaRequestV2.newBuilder()
            .setConfigData(this.configData)
            .setDisplayName(this.displayName)
            .setSchemaType(this.schemaType)
            .setVersion(this.version)
            .setDescription(this.description)
            .addAllStringAttributes(this.stringAttributes)
            .addAllIntegerAttributes(this.integerAttributes)
            .addAllDecimalAttributes(this.decimalAttributes)
            .addAllBooleanAttributes(this.booleanAttributes)
            .addAllDateAttributes(this.dateAttributes)
            .addAllDatetimeAttributes(this.datetimeAttributes)
            .addAllStringEnumAttributes(this.stringEnumAttributes)
            .addAllIntegerEnumAttributes(this.integerEnumAttributes)
            .addAllDecimalEnumAttributes(this.decimalEnumAttributes)
            .build();

    IdentityV2.BuildSchemaResponseV2 response = bridge.getIdentityV2().buildSchema(req);

    if (response.getError() != Shared.Error.getDefaultInstance()) {
      throw new Exception(response.getError().getMessage());
    }

    return Schema.fromProto(response.getSchema());
  }
}
